### PR TITLE
Dev

### PR DESCRIPTION
**Pull request by Tuxman**
GraalVM update with new API routes, key management, and structure for Valkyrie functions as the candidate for release 3.0

### CODE REVIEW

File: src/main/scala/bifrost/api/http/AssetApiRoute.scala [hunk 1]
```diff
@@ -42,20 +45,22 @@ case class AssetApiRoute (override val settings: Settings, nodeViewHolderRef: Ac
...
               reqId = id
               require((json \\ "jsonrpc").head.asString.get == "2.0")
               val params = (json \\ "params").head.asArray.get
               require(params.size <= 5, s"size of params is ${params.size}")
 
               (json \\ "method").head.asString.get match {
-                case "redeemAssets" => redeemAssets(view, params.head, id).asJson
-                case "transferAssets" => transferAssets(view, params.head, id).asJson
-                case "createAssets" => createAssets(view, params.head, id).asJson
+                case "redeemAssets" => redeemAssets(params.head, id)
```
Comment (AHBruns) [inline src/main/scala/bifrost/api/http/AssetApiRoute.scala]: should these be .asJson ?

File: src/main/scala/bifrost/contract/ValkyrieFunctions.scala [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+          source match {
+            case "assetCreated" => CompilerDirectives.transferToInterpreter(); throw context.createUnwind("ca")
+            case "assetTransferred" => CompilerDirectives.transferToInterpreter(); throw context.createUnwind("ta")
+            case "polyTransferred" => CompilerDirectives.transferToInterpreter(); throw context.createUnwind("tp")
+          }
+        }
+
+        override def onReturnExceptional(context: EventContext, frame: VirtualFrame, exception: Throwable): Unit = ???
+
+        import com.oracle.truffle.api.frame.VirtualFrame
```
Comment (AHBruns) [inline src/main/scala/bifrost/contract/ValkyrieFunctions.scala]: should probably hoist this to the top of the file.

File: src/main/scala/bifrost/contract/ValkyrieFunctions.scala [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+
+        override def onUnwind(context: EventContext, frame: VirtualFrame, info: Object): Object = {
+          info match {
```
Comment (AHBruns) [inline src/main/scala/bifrost/contract/ValkyrieFunctions.scala]: should have default case to make later debugging easier. Something like, log("Valkyrie.onUnwind unmatched info arg").

File: src/main/scala/bifrost/scorexMod/GenericNodeViewHolder.scala [hunk 4]
```diff
@@ -240,8 +240,12 @@ trait GenericNodeViewHolder[T, P <: Proposition, TX <: GenericBoxTransaction[P,
       val theyAreYounger = comparison == HistoryComparisonResult.Younger
       val notSendingBlocks = extensionOpt.isEmpty
 
-      if(notSendingBlocks && theyAreYounger) throw new Exception("Other node was younger but we didn't have blocks to send")
-      
+      //if(notSendingBlocks && theyAreYounger) throw new Exception("Other node was younger but we didn't have blocks to send")
+
+      if(notSendingBlocks && theyAreYounger) {
+        log.debug(s"Unable to sync without common ancestor")
```
Comment (AHBruns) [inline src/main/scala/bifrost/scorexMod/GenericNodeViewHolder.scala]: could be more verbose.

File: src/main/scala/bifrost/transaction/BifrostTransactionCompanion.scala [hunk 5]
```diff
@@ -421,13 +428,17 @@ object ContractCompletionCompanion extends Serializer[ContractCompletion] {
...
+      ContractTransactionCompanion.commonToBytes(cc),
+      cc.data.getBytes,
+      Ints.toByteArray(cc.data.getBytes.length)
     )
   }
 
   //noinspection ScalaStyle
   override def parseBytes(bytes: Array[Byte]): Try[ContractCompletion] = Try {
-
+    val dataLen: Int = Ints.fromByteArray(bytes.slice(bytes.length - Ints.BYTES, bytes.length))
```
Comment (AHBruns) [inline src/main/scala/bifrost/transaction/BifrostTransactionCompanion.scala]: Bring up at stand up. Not fully clear how arbitrary data field gets serialized.

File: src/main/scala/bifrost/wallet/PolyTransferGenerator.scala [hunk 6]
```diff
@@ -53,6 +53,6 @@ object PolyTransferGenerator {
       }.toSeq
 
     val recipient = pubkeys(Random.nextInt(pubkeys.size))
-    PolyTransfer.create(wallet, IndexedSeq((recipient, Random.nextInt(100))), Random.nextInt(100))
+    PolyTransfer.create(wallet, IndexedSeq((recipient, Random.nextInt(100))), Random.nextInt(100), "")
```
Comment (AHBruns) [inline src/main/scala/bifrost/wallet/PolyTransferGenerator.scala]: We might want to overload the transfer functions with versions that accept no arbitrary data field and instead add an empty string themselves. Will help code clarity.

File: src/test/scala/bifrost/transaction/ContractTransactionSpec.scala [hunk 7]
```diff
@@ -186,7 +190,9 @@ class ContractTransactionSpec extends PropSpec
         ++ Longs.toByteArray(timestamp)
         ++ fees.flatMap { case (prop, feeValue) => prop.pubKeyBytes ++ Longs.toByteArray(feeValue) })
 
-    val messageToSign = FastCryptographicHash(contractBox.value.noSpaces.getBytes ++ hashNoNonces)
+    val messageToSign = Bytes.concat(
```
Comment (AHBruns) [inline src/test/scala/bifrost/transaction/ContractTransactionSpec.scala]: Should make this pattern a util function to improve readability.

File: src/main/scala/bifrost/contract/ValkyrieFunctions.scala (hunk 2 above)
Comment (Tuxman) [reply src/main/scala/bifrost/contract/ValkyrieFunctions.scala]: For whatever reason that import needs to be there. If moved out of the object, will error.

File: src/test/scala/bifrost/transaction/ContractTransactionSpec.scala (hunk 7 above)
Comment (Tuxman) [reply src/test/scala/bifrost/transaction/ContractTransactionSpec.scala]: Can definitely make improvements on this, something to throw into the backlog.

File: src/main/scala/bifrost/scorexMod/GenericNodeViewHolder.scala (hunk 4 above)
Comment (Tuxman) [reply src/main/scala/bifrost/scorexMod/GenericNodeViewHolder.scala]: "Error: Trying to sync local node with remote node. Failed to find common ancestor within block history. Check that you are attempting to sync to the correct version of the blockchain."

File: src/main/scala/bifrost/wallet/PolyTransferGenerator.scala (hunk 6 above)
Comment (Tuxman) [reply src/main/scala/bifrost/wallet/PolyTransferGenerator.scala]: The current implementation of transferPolys requires a "data" field in the API request params, can change spec to allow for optional parameters in the API.

File: src/main/scala/bifrost/BifrostApp.scala [hunk 8]
```diff
@@ -45,6 +50,7 @@ class BifrostApp(val settingsFilename: String) extends GenericApplication with R
     WalletApiRoute(settings, nodeViewHolderRef),
     ContractApiRoute(settings, nodeViewHolderRef, networkController),
     AssetApiRoute(settings, nodeViewHolderRef),
+    WalletApiRouteRPC(settings, nodeViewHolderRef),
```
[the remark this answers is not in the document]
Comment (Tuxman) [reply src/main/scala/bifrost/BifrostApp.scala]: Yes, but make sure it does not break things when removing.

File: src/main/scala/bifrost/BifrostApp.scala [hunk 9]
```diff
@@ -17,6 +17,11 @@ import scorex.core.network.message.MessageSpec
 import scorex.core.transaction.box.proposition.ProofOfKnowledgeProposition
 import scorex.core.transaction.state.PrivateKey25519
 
+import org.graalvm.polyglot.Context
+import java.lang.management.ManagementFactory
+import com.sun.management.HotSpotDiagnosticMXBean
```
[the remark this answers is not in the document]
Comment (Tuxman) [reply src/main/scala/bifrost/BifrostApp.scala]: Not worth taking out at this time.

File: src/main/scala/bifrost/api/http/AssetApiRoute.scala [hunk 10]
```diff
@@ -66,54 +71,81 @@ case class AssetApiRoute (override val settings: Settings, nodeViewHolderRef: Ac
...
-    }
+      val tempTx = params.as[AssetRedemption] match {
+        case Right(a: AssetRedemption) => a
+        case Left(e) => throw new JsonParsingException(s"Could not parse AssetRedemption: $e")
+      }
 
-    val realSignature = PrivateKey25519Companion.sign(selectedSecret, tempTx.messageToSign)
-    val modifiedSignatures = tempTx.signatures.map { case (key, fakeSigs) =>
-      (key, fakeSigs.map(_ => realSignature))
+      val realSignature = PrivateKey25519Companion.sign(selectedSecret, tempTx.messageToSign)
```
[the remark this answers is not in the document]
Comment (Tuxman) [reply src/main/scala/bifrost/api/http/AssetApiRoute.scala]: Something to go over from a cryptographic standpoint at a later date.

File: src/main/scala/bifrost/api/http/AssetApiRoute.scala [hunk 11]
```diff
@@ -66,54 +71,81 @@ case class AssetApiRoute (override val settings: Settings, nodeViewHolderRef: Ac
...
-    val fee: Long = (params \\ "fee").head.asNumber.flatMap(_.toLong).getOrElse(0L)
-    val hub = PublicKey25519Proposition(Base58.decode((params \\ "hub").head.asString.get).get)
-    val assetCode: String = (params \\ "assetCode").head.asString.getOrElse("")
-    val data: String = (params \\ "data").head.asString.getOrElse("")
-    val tx = AssetTransfer.create(wallet, IndexedSeq((recipient, amount)), fee, hub, assetCode, data).get
-    nodeViewHolderRef ! LocallyGeneratedTransaction[ProofOfKnowledgeProposition[PrivateKey25519], AssetTransfer](tx)
-    tx.json
+  private def transferAssets(params: Json, id: String): Future[Json] = {
+
+    println(s">>>>>>>> transferAssets: $params")
```
[the remark this answers is not in the document]
Comment (Tuxman) [reply src/main/scala/bifrost/api/http/AssetApiRoute.scala]: Yeah that's fine.

File: src/main/scala/bifrost/contract/modules/BaseModuleWrapper.scala [hunk 12]
```diff
@@ -124,7 +134,6 @@ object BaseModuleWrapper {
     val jsre: NashornScriptEngine = new NashornScriptEngineFactory().getScriptEngine.asInstanceOf[NashornScriptEngine]
```
[the remark this answers is not in the document]
Comment (Tuxman) [reply src/main/scala/bifrost/contract/modules/BaseModuleWrapper.scala]: Yes, the way we are using Esprima works better with Nashorn. I am removing Esprima though, and with it Nashorn entirely.

**Pull request merged**